### PR TITLE
schema cleanup: drop network_total_transactions from checkpoints

### PR DESCRIPTION
**Pull request by gegaowp**
## Description 

title, more contexts can be found in https://linear.app/mysten-labs/issue/DP-63/clean-up-network-total-transactions-vs-max-tx-sequence-number-and-min

## Test plan 

ci to ensure no references to this column any more

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:
- [ ] REST API:

### CODE REVIEW

File: crates/sui-indexer/src/store/pg_indexer_store.rs [hunk 1]
```diff
@@ -1131,17 +1131,20 @@ impl PgIndexerStore {
...
-                                .map(|o| o.unwrap_or(0))?;
-
-                            result as u64
+                                .await?
+                                .ok_or_else(|| {
+                                    IndexerError::PersistentStorageDataCorruptionError(
+                                        "max_tx_sequence_number should not be None".to_string(),
+                                    )
+                                })?;
+                            (last_epoch_max_tx_sequence_number as u64).saturating_add(1)
```
Comment (gegaowp) [inline crates/sui-indexer/src/store/pg_indexer_store.rs:1145]: having var `last_epoch_max_tx_sequence_number` to make conversion from `last_epoch_max_tx_sequence_number` to `network_total_transactions` a bit more obvious

File: crates/sui-indexer/src/store/pg_indexer_store.rs [hunk 2]
```diff
@@ -1458,16 +1461,18 @@ impl PgIndexerStore {
...
-            .context("Failed to get network total transactions in epoch")
-            .map(|v| v as u64)
+            .first::<Option<i64>>(&mut connection)
+            .await?
+            .ok_or_else(|| {
+                IndexerError::PersistentStorageDataCorruptionError(
+                    "max_tx_sequence_number should not be None".to_string(),
+                )
+            })?;
+        Ok((max_tx_sequence_number as u64).saturating_add(1))
```
Comment (gegaowp) [inline crates/sui-indexer/src/store/pg_indexer_store.rs:1475]: same as the other

File: crates/sui-indexer/migrations/pg/2024-09-25-211901_drop_network_total_transactions/down.sql [hunk 3]
```diff
@@ -0,0 +1 @@
+ALTER TABLE checkpoints ADD COLUMN network_total_transactions BIGINT NOT NULL DEFAULT 0;
```
[the remark this answers is not in the document]
Comment (bmwill) [reply crates/sui-indexer/migrations/pg/2024-09-25-211901_drop_network_total_transactions/down.sql:1]: I think i'm missing some context but the PR title says we're removing `network_total_transactions` but this migration adds this column?

File: crates/sui-indexer/src/models/checkpoints.rs [hunk 4]
```diff
@@ -118,6 +116,13 @@ impl TryFrom<StoredCheckpoint> for RpcCheckpoint {
                 })
                 .collect::<Result<Vec<TransactionDigest>, IndexerError>>()?
         };
+
+        let max_tx_sequence_number = checkpoint.max_tx_sequence_number.ok_or_else(|| {
+            IndexerError::PersistentStorageDataCorruptionError(
+                "max_tx_sequence_number should not be None".to_string(),
```
[the remark this answers is not in the document]
Comment (gegaowp) [reply crates/sui-indexer/src/models/checkpoints.rs:122]: @amnn good callout regarding empty checkpoint!

> I notice a lot of adding +1/-1 

indeed this is mainly b/c both upstream checkpoint data and JSON RPC checkpoint need `network_total_transactions` while we only stored `max_tx_sequence_number`, which less than `network_total_transactions` by 1.

```
tx_lo = 1 + max(tx for checkpoint c < C for tx in c)
tx_hi = 1 + max(tx for checkpoint c <= C for tx in c) 
```

this would work indeed, while I feel it's a bit hard to grasp the meaning of tx_lo & tx_hi intuitively, another alternative that would also cover the empty checkpoint case would be

- network_total_transactions for tx count at the end of the checkpoint, this is also in checkpoint;
- checkpoint_total_transactions for tx count of current checkpoint, which is 0 for empty case

and then we can derive min & max, now mainly for pruning and change the meaning like to half-open orthogonally if we want, wdyt?

File: crates/sui-indexer/migrations/pg/2024-09-25-211901_drop_network_total_transactions/down.sql (hunk 3 above)
[the remark this answers is not in the document]
Comment (gegaowp) [reply crates/sui-indexer/migrations/pg/2024-09-25-211901_drop_network_total_transactions/down.sql:1]: this is the down.sql so it's to revert the dropping.

[the remark this answers is not in the document]
Comment (gegaowp) [reply crates/sui-indexer/migrations/pg/2024-09-25-211901_drop_network_total_transactions/down.sql:1]: although as Ashok pointed out, the empty checkpoint can be an issue following the current approach, other alternatives might change the migration here

[the remark this answers is not in the document]
Comment (gegaowp) [reply crates/sui-indexer/migrations/pg/2024-09-25-211901_drop_network_total_transactions/down.sql:1]: re adding back NOT NULL with a fake value <> adding back NULL-able column, I thought it twice and feel it's prob safer to have a slightly relaxed reverse migration by adding back NULL-able column to avoid mixing fake values and real populated values.

File: crates/sui-indexer/src/models/checkpoints.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (amnn) [reply crates/sui-indexer/src/models/checkpoints.rs:122]: The counts idea is interesting -- I quite like it -- the only thing to check is whether we ever need to filter on `min_tx_sequence_number`, because if so, you want to store that directly, rather than compute it (so that you can build an index over it).

FWIW, I think what makes the `tx_lo`/`tx_hi` description non-intuitive is trying to make it correct in all cases (particularly the empty case), but there is a fairly natural intuitive description of it, which is that what we're doing is that we are tracking the boundaries between checkpoints always (with the upperbound of one checkpoint equalling the lowerbound of the next).

Thinking in those terms, using half open intervals is the easiest way to capture all the edge cases, because it can represent an empty checkpoint while still remembering where in the stream of checkpoints/transactions it belongs.

[the remark this answers is not in the document]
Comment (gegaowp) [reply crates/sui-indexer/src/models/checkpoints.rs:122]: > whether we ever need to filter on min_tx_sequence_number

I like the forward thinking! although I think for the tx seq -> cp seq lookup, we will want to leverage transaction tables instead of building an index on checkpoint table. (also the case when people have a tx seq, knowing that it's a min_tx_sequence_number and want to find the related cp seems pretty rare to me, this may be debatable), does that make sense?

[the remark this answers is not in the document]
Comment (amnn) [reply crates/sui-indexer/src/models/checkpoints.rs:122]: > also the case when people have a tx seq, knowing that it's a min_tx_sequence_number and want to find the related cp seems pretty rare to me, this may be debatable

The kind of query that I'm talking about is not one where the tx you hold in your hand happens to be a min tx sequence number, but one where you wanted to select checkpoints but by applying a bound on their min_tx_sequence_number.

This is not that uncommon -- we do this on the `epochs` table, bounding by `first_checkpoint_id` (for consistent read reasons). I think the broader point is that there is value in avoiding computing the actual bounds, because they are more likely to show up in filters, as opposed to the actual count of transactions, which is unlikely to show up in filters.

[the remark this answers is not in the document]
Comment (gegaowp) [reply crates/sui-indexer/src/models/checkpoints.rs:122]: the broader point makes sense, will do it that way.

File: crates/sui-graphql-rpc/src/types/checkpoint.rs [hunk 5]
```diff
@@ -117,8 +117,9 @@ impl Checkpoint {
     }
 
     /// The total number of transaction blocks in the network by the end of this checkpoint.
-    async fn network_total_transactions(&self) -> Option<UInt53> {
-        Some(self.network_total_transactions_impl().into())
+    async fn network_total_transactions(&self) -> Result<Option<UInt53>> {
```
Comment (gegaowp) [inline crates/sui-graphql-rpc/src/types/checkpoint.rs:120]: @amnn changing the return type here, any wisdom? I suppose we will need to handle this everywhere when we remove NOT NULL contraints for those in fact NOT NULL columns.

File: crates/sui-indexer/migrations/pg/2024-09-25-211901_drop_network_total_transactions/up.sql [hunk 6]
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE checkpoints DROP COLUMN network_total_transactions;
+UPDATE checkpoints SET max_tx_sequence_number = max_tx_sequence_number + 1 WHERE max_tx_sequence_number IS NOT NULL;
```
Comment (gegaowp) [inline crates/sui-indexer/migrations/pg/2024-09-25-211901_drop_network_total_transactions/up.sql:2]: this might take mins to finish, I can split this change into steps
- adding a new column like `new_max_tx_sequence_number` and populate both
- backfill `new_max_tx_sequence_number`
- remove `max_tx_sequence_number` and rename `new_max_tx_sequence_number` to `max_tx_sequence_number`

it's a bit wasteful .. but I am okay to do that if needed. cc @lxfind

Comment (lxfind) [inline crates/sui-indexer/migrations/pg/2024-09-25-211901_drop_network_total_transactions/up.sql:2]: I don't actually understand. Why + 1?

Comment (gegaowp) [reply crates/sui-indexer/migrations/pg/2024-09-25-211901_drop_network_total_transactions/up.sql:2]: Ashok proposed a slightly different semantics for min and max seq numbers as

> There is an alternative, which is to store (for checkpoint C):
tx_lo = 1 + max(tx for checkpoint c < C for tx in c)
tx_hi = 1 + max(tx for checkpoint c <= C for tx in c) 
These are guaranteed to exist, even for empty checkpoints, so we can use a NOT NULL constraint if we wanted to.
You can compute network_total_transactions from them (it's tx_hi if transactions are 0-indexed, or tx_hi - 1 otherwise).
The number of transactions in C is always tx_hi - tx_lo.

Basically that changes the max_tx_sequence_number from inclusive to exclusive, hence the +1 for the semantics change.

Comment (lxfind) [reply crates/sui-indexer/migrations/pg/2024-09-25-211901_drop_network_total_transactions/up.sql:2]: Are there empty checkpoints?
If we do want to go that route, I think max_tx_sequence_number would be a very confusing name.

Comment (lxfind) [reply crates/sui-indexer/migrations/pg/2024-09-25-211901_drop_network_total_transactions/up.sql:2]: It seems `network_total_transactions` would have a much clearer semantics than the new `max_tx_sequence_number`

Comment (gegaowp) [reply crates/sui-indexer/migrations/pg/2024-09-25-211901_drop_network_total_transactions/up.sql:2]: > Are there empty checkpoints?

no empty checkpoint yet.

> If we do want to go that route, I think max_tx_sequence_number would be a very confusing name

this sounds familiar :) -- I actually proposed `network_total_transactions` & `checkpoint_total_transaction` instead of min & `max_tx_sequence_number` in a thread above with @amnn, I think that's more intuitive indeed https://github.com/MystenLabs/sui/pull/19560#discussion_r1776856973 I saw Ashok's broader point about pre-computation and filtering too though hence changing it that way

Comment (amnn) [reply crates/sui-indexer/migrations/pg/2024-09-25-211901_drop_network_total_transactions/up.sql:2]: How about if you dropped `max_tx_sequence_number` and renamed `network_total_transactions`? (modulo the conversation on what the name should be, below)

File: crates/sui-indexer/src/store/pg_indexer_store.rs (hunk 1 above)
Comment (amnn) [reply crates/sui-indexer/src/store/pg_indexer_store.rs:1145]: Do you still need the `+1` here?

File: crates/sui-indexer/src/store/pg_indexer_store.rs (hunk 2 above)
Comment (amnn) [reply crates/sui-indexer/src/store/pg_indexer_store.rs:1475]: ditto

**Review by amnn — COMMENTED** [review]
I'm in two minds about whether these columns should be nullable. I know we spoke about generally making columns nullable when we are adding them, but I wonder if we can avoid that issue by renaming existing columns in this particular case.

File: crates/sui-indexer/migrations/pg/2024-09-25-211901_drop_network_total_transactions/up.sql (hunk 6 above)
Comment (gegaowp) [reply crates/sui-indexer/migrations/pg/2024-09-25-211901_drop_network_total_transactions/up.sql:2]: hmm min&max <> lo&hi are sorta the same to me, the names per se do not indicate they are inclusive or exclusive -- it's actually good if tx_hi is less confusing and happy to do it that way

Comment (gegaowp) [reply crates/sui-indexer/migrations/pg/2024-09-25-211901_drop_network_total_transactions/up.sql:2]: great idea, thanks!

File: crates/sui-graphql-rpc/src/types/checkpoint.rs [hunk 7]
```diff
@@ -204,8 +205,11 @@ impl Checkpoint {
...
     }
 
-    pub(crate) fn network_total_transactions_impl(&self) -> u64 {
-        self.stored.network_total_transactions as u64
+    pub(crate) fn network_total_transactions_impl(&self) -> Result<u64, Error> {
+        let total_tx = self.stored.max_tx_sequence_number.ok_or(Error::Internal(
+            "Checkpoint max_tx_sequence_number should not be null".to_string(),
+        ))?;
+        Ok(total_tx as u64)
     }
```
[the remark this answers is not in the document]
Comment (gegaowp) [reply crates/sui-graphql-rpc/src/types/checkpoint.rs:213]: I thought the nullable move was for migrations and scheme changes, and on the reader end, if it's a in-fact-non-null column, we should fail fast and error whenever we found a null?

File: crates/sui-indexer/migrations/pg/2024-09-25-211901_drop_network_total_transactions/up.sql (hunk 6 above)
Comment (amnn) [reply crates/sui-indexer/migrations/pg/2024-09-25-211901_drop_network_total_transactions/up.sql:2]: There is a difference: `min` and `max` are definitely inclusive bounds, whereas `lo` and `hi` do not have that association, so people are more likely to read the docs to confirm whether the bound is exclusive or not and less likely to be confused by an exclusive `hi` vs an exclusive `max`. What's more, these terms are already used this way in the GraphQL codebase that will read these tables so there will also be no confusion there (and we document the convention there too).

File: crates/sui-graphql-rpc/src/types/checkpoint.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (amnn) [reply crates/sui-graphql-rpc/src/types/checkpoint.rs:213]: > I thought the nullable move was for migrations and scheme changes

It is, in the initial instance, but also note my earlier comment:

> We do try and make fields in the GraphQL schema nullable/optional if we think they might go away because of deprecations, but because we haven't been forced to do that by the types, we haven't done it everywhere, so making the columns nullable will make sure we do that thoroughly.

This will not be possible if we insulate the GraphQL schema from the nullability of the DB schema by failing fast, and then re-wrapping it in an `Option` (the types will not be forcing us in that case), and in the end, if we do deprecate a field, we would then need to remember to undo the unwrap/rewrap to prevent the deprecation from breaking existing queries because of the fail fast logic.
